cmd/scale.go

```diff
@@ -24,9 +29,105 @@ func newScaleCommand() *Command {
 	vmCmd := BuildCommand(cmd, runScaleVM, vmCmdStrings.Usage, vmCmdStrings.Short, vmCmdStrings.Long, os.Stdout, requireSession, requireAppName)
 	vmCmd.Args = cobra.RangeArgs(0, 1)
 
+	balanceCmdStrings := docstrings.Get("scale.balanced")
+	balanceCmd := BuildCommand(cmd, runBalanceScale, balanceCmdStrings.Usage, balanceCmdStrings.Short, balanceCmdStrings.Long, os.Stdout, requireSession, requireAppName)
+	balanceCmd.Args = cobra.RangeArgs(0, 2)
+
+	standardCmdStrings := docstrings.Get("scale.standard")
+	standardCmd := BuildCommand(cmd, runStandardScale, standardCmdStrings.Usage, standardCmdStrings.Short, standardCmdStrings.Long, os.Stdout, requireSession, requireAppName)
+	standardCmd.Args = cobra.RangeArgs(0, 2)
+
+	setCmdStrings := docstrings.Get("scale.set")
+	setCmd := BuildCommand(cmd, runSetParamsOnly, setCmdStrings.Usage, setCmdStrings.Short, setCmdStrings.Long, os.Stdout, requireSession, requireAppName)
+	setCmd.Args = cobra.RangeArgs(0, 2)
+
+	showCmdStrings := docstrings.Get("scale.show")
+	BuildCommand(cmd, runShow, showCmdStrings.Usage, showCmdStrings.Short, showCmdStrings.Long, os.Stdout, requireSession, requireAppName)
+
 	return cmd
 }
 
+func runBalanceScale(ctx *CmdContext) error {
+	return actualScale(ctx, true, false)
+}
+
+func runStandardScale(ctx *CmdContext) error {
+	return actualScale(ctx, false, false)
+}
+
+func runSetParamsOnly(ctx *CmdContext) error {
+	return actualScale(ctx, false, true)
+}
+
+func actualScale(ctx *CmdContext, balanceRegions bool, setParamsOnly bool) error {
+	currentcfg, err := ctx.Client.API().AppAutoscalingConfig(ctx.AppName)
+
+	newcfg := api.UpdateAutoscaleConfigInput{AppID: ctx.AppName}
+
+	newcfg.BalanceRegions = &balanceRegions
+	newcfg.MinCount = &currentcfg.MinCount
+	newcfg.MaxCount = &currentcfg.MaxCount
+
+	kvargs := make(map[string]string)
+
+	for _, pair := range ctx.Args {
+		parts := strings.SplitN(pair, "=", 2)
+		if len(parts) != 2 {
+			return fmt.Errorf("Secrets must be provided as NAME=VALUE pairs (%s is invalid)", pair)
+		}
+		key := parts[0]
+		value := parts[1]
+		kvargs[strings.ToLower(key)] = value
+	}
+
+	minval, found := kvargs["min"]
+
+	if found {
+		minint64val, err := strconv.ParseInt(minval, 10, 64)
+
+		if err != nil {
+			return errors.New("could not parse min count value")
+		}
+		minintval := int(minint64val)
+		newcfg.MinCount = &minintval
+		delete(kvargs, "min")
+	}
+
+	maxval, found := kvargs["max"]
+
+	if found {
+		maxint64val, err := strconv.ParseInt(maxval, 10, 64)
+
+		if err != nil {
+			return errors.New("could not parse max count value")
+		}
+		maxintval := int(maxint64val)
+		newcfg.MaxCount = &maxintval
+		delete(kvargs, "max")
+	}
+
+	if len(kvargs) != 0 {
+		unusedkeys := ""
+		for k := range kvargs {
+			if unusedkeys == "" {
+				unusedkeys = k
+			} else {
+				unusedkeys = unusedkeys + ", " + k
+			}
+		}
+		return errors.New("unrecognised parameters in command:" + unusedkeys)
```

Comment:
You can reduce this block to `strings.Join(unusedKeys, ", ")`

Reply:
Except that the source for the unusedKeys strings is coming from the map keys and strings.Join doesn't take a map and the process of making an array beforehand is just as wordy. There's some trickery that can be done with reflect but its not pretty. I've gone for simple in this case.